### PR TITLE
Add tensor step to RMSprop, Rprop, and Adadelta

### LINKED ISSUE #115679: Dynamo'ing Rprop, RMSprop, and Adadelta misses incrementing step due to skipping _init_group
### 🐛 Describe the bug

Currently the step update for Rprop, RMSprop and Adadelta occurs in _init_group, which gets skipped in dynamo. This means that running torch.compile on these optimizers will correctly update the parameters but not the step.

_init_group is currently disabled in dynamo today due to slow tracing over loops. However, it seems like a bug that _init_group is skipped entirely in execution too.

And yes, moving the step update to the actual `step` function would technically "solve" this issue, but I do not believe it is the right fix. We should still run _init_group before step even after dynamo'ing. cc @vincentqb @jbschlosser @albanD @crcrpar @ezyang @msaroufim @wconstab @bdhirsh @anijain2305 @zou3519 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng @mlazos

### Error logs

```
(pytorch-3.10) [janeyx@devgpu023.odn1 ~/local/pytorch (e9184613)]$ python playground.py
expected in eager:
step 1
step 2
step 3
step 3
step 4
step 5
what actually happens after dynamo:
step 1
step 1
step 1
step 1
step 2
step 3
```

You may be curious why the steps suddenly start incrementing after a bit. That is because recompilation due to the .grad fields becoming all None causes dynamo to run again, which will correctly call `_init_group`.

### Minified repro

```
import torch
def training_loop():
    input = torch.tensor(
        [0.1, 0.2, 0.3, 0.4, 0.5, 0.6]
    ).reshape(3, 2)

    model = torch.nn.Sequential(
        torch.nn.Linear(2, 3),
        torch.nn.Sigmoid(),
        torch.nn.Linear(3, 1),
        torch.nn.Sigmoid(),
    )

    params = list(model.parameters())
    optimizer = torch.optim.Rprop(params)

    for i in range(6):
        optimizer.zero_grad()
        # Test that step behaves as expected (a no-op) when grads are set to None
        if i != 3:
            output = model(input)
            loss = output.sum()
            loss.backward()

        optimizer.step()
        print("step", optimizer.state[params[0]]["step"])

compiled_training_loop = torch._dynamo.optimize("eager")(training_loop)

print("expected in eager:")
training_loop()

print("what actually happens after dynamo:")
compiled_training_loop()
```

### Versions

on main

### PR DESCRIPTION
**Pull request by mlazos**
Fixes #115679

Creates a tensor step, and increments it in the compute portion of the optimizers.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/optim/adadelta.py [hunk 1]
```diff
@@ -63,7 +81,9 @@ def _init_group(self, group, params_with_grad, grads, square_avgs, acc_deltas):
 
             # Lazy state initialization
             if len(state) == 0:
-                state["step"] = 0
+                state["step"] = torch.zeros(
+                    (), device=p.device, dtype=_get_scalar_dtype()
```
Comment (janeyx99) [inline torch/optim/adadelta.py:85]: We don't want this on p's device by default on eager

File: torch/optim/adadelta.py [hunk 2]
```diff
@@ -198,7 +226,9 @@ def adadelta(
     """
     # We still respect when the user inputs False for foreach.
     if foreach is None:
-        _, foreach = _default_to_fused_or_foreach(params, differentiable, use_fused=False)
+        _, foreach = _default_to_fused_or_foreach(
```
Comment (janeyx99) [inline torch/optim/adadelta.py:229]: What linter are you using and why is it relinting already linted lines?

File: torch/optim/adadelta.py [hunk 3]
```diff
@@ -297,10 +342,15 @@ def _multi_tensor_adadelta(
             if maximize:
                 torch._foreach_add_(device_grads, device_params, alpha=weight_decay)
             else:
-                device_grads = torch._foreach_add(device_grads, device_params, alpha=weight_decay)
+                device_grads = torch._foreach_add(
+                    device_grads, device_params, alpha=weight_decay
+                )
 
+        torch._foreach_add_(steps, 1)
```
Comment (janeyx99) [inline torch/optim/adadelta.py:349]: Check the adam.py code, there is a heuristic we have for steps on CPU.

File: torch/optim/adadelta.py (hunk 1 above)
Comment (mlazos) [reply torch/optim/adadelta.py:85]: oh, should I add a capturable arg here then?

File: torch/optim/adadelta.py (hunk 2 above)
Comment (mlazos) [reply torch/optim/adadelta.py:229]: lintrunner on my devgpu? haha, not sure

**Review by janeyx99 — COMMENTED** [review]
Thanks for taking this on! I looked over one optimizer currently, and stopped reviewing because I noticed the following for all the impls. step needs to be on CPU for performance reasons on eager to avoid launching CUDA/XLA kernels (which are expensive).

If changing it to be on CPU is prohibitive for compilation, we will need to expose public capturable APIs for all the optimizers (with docs and testing) and gate moving step to the p.device only if capturable. This will be better to do in 3 separate PRs if required.

File: torch/optim/adadelta.py (hunk 2 above)
Comment (janeyx99) [reply torch/optim/adadelta.py:229]: nah, it looks like your max line width is too narrow

**Comment by mlazos** [comment]
Closing in favor of separate PRs

**Pull request closed without merging**